crates/ide_completion/src/patterns.rs

```diff
@@ -27,6 +27,14 @@ pub(crate) enum ImmediatePrevSibling {
     Attribute,
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) enum TypeAnnotation {
+    Let(Option<ast::Pat>),
+    FnParam(Option<ast::Pat>),
+    RetType(Option<ast::Expr>),
```

Comment:
And what is the problem of return types (missing body node in original file)?

```suggestion
    RetType(Option<ast::Ty>),
```

Reply:
Em, here I need the body expression of the function, then infer its type, so I need a `ast::Expr`?

Reply:
oh sorry I misunderstood